src/script/sign.h

```diff
@@ -68,6 +68,392 @@ struct SignatureData {
+            WriteCompactSize(s, entry.second.size());
+            s << MakeSpan(entry.second);
+        }
+        // Write any hd keypaths
+        for (auto keypath_pair : hd_keypaths) {
+            s << SerializeToVector(PSBT_BIP32_KEYPATH_SIGHASH, Span<const unsigned char>(keypath_pair.first.begin(), keypath_pair.first.size()));
+            WriteCompactSize(s, keypath_pair.second.size() * sizeof(uint32_t));
+            for (auto& key : keypath_pair.second) {
+                s << key;
+            }
+        }
+
+        // Write the number of inputs
+        if (num_ins > 0) {
+            s << PSBT_NUM_IN_VIN;
+            WriteCompactSize(s, num_ins);
+        }
+
+        // Write the unknown things
+        for(auto& entry : unknown) {
+            s << entry.first;
+            s << entry.second;
+        }
+
+        // Separator
+        s << PSBT_SEPARATOR;
+
+        for (unsigned int i = 0; i < tx.vin.size(); i++) {
+            CTxIn in = tx.vin[i];
+            PartiallySignedInput psbt_in = inputs[i];
+            if (in.scriptSig.empty() && in.scriptWitness.IsNull()) {
+                // Write the utxo
+                // If there is a non-witness utxo, then don't add the witness one.
+                if (psbt_in.non_witness_utxo) {
+                    WriteCompactSize(s, sizeof(PSBT_UNSIGNED_TX_NON_WITNESS_UTXO));
+                    s << PSBT_UNSIGNED_TX_NON_WITNESS_UTXO;
+                    WriteCompactSize(s, ::GetSerializeSize(psbt_in.non_witness_utxo, SER_NETWORK, PROTOCOL_VERSION));
+                    s << psbt_in.non_witness_utxo;
+                } else if (!psbt_in.witness_utxo.IsNull()) {
+                    WriteCompactSize(s, sizeof(PSBT_REDEEMSCRIPT_WITNESS_UTXO));
+                    s << PSBT_REDEEMSCRIPT_WITNESS_UTXO;
+                    WriteCompactSize(s, ::GetSerializeSize(psbt_in.witness_utxo, SER_NETWORK, PROTOCOL_VERSION));
+                    s << psbt_in.witness_utxo;
+                }
+
+                // Write any partial signatures
+                for (auto sig_pair : psbt_in.partial_sigs) {
+                    s << SerializeToVector(PSBT_WITNESSSCRIPT_PARTIAL_SIG, Span<const unsigned char>(sig_pair.first.begin(), sig_pair.first.size()));
+                    WriteCompactSize(s, sig_pair.second.size());
+                    s << MakeSpan(sig_pair.second);
+                }
+
+                // Write the sighash type
+                if (psbt_in.sighash_type > 0) {
+                    WriteCompactSize(s, sizeof(PSBT_BIP32_KEYPATH_SIGHASH));
+                    s << PSBT_BIP32_KEYPATH_SIGHASH;
+                    WriteCompactSize(s, sizeof(psbt_in.sighash_type));
+                    s << psbt_in.sighash_type;
+                }
+
+                // Write the index
+                if (use_in_index) {
+                    WriteCompactSize(s, sizeof(PSBT_NUM_IN_VIN));
+                    s << PSBT_NUM_IN_VIN;
+                    WriteCompactSize(s, sizeof(psbt_in.index));
```

Comment:
> This is a lot easier if you'd have a single-argument SerializeToVector first (see above).

Above where?

I'm not necessarily sure that that will always work due to the different data types.

> Perhaps it's generally better to use the separator approach instead?

I think it generally is better to use the separator approach. By default, this code does not use input indexes unless a psbt that uses input indexes was provided. That option was provided at the suggestion of someone else in order to handle very large transactions where almost all inputs were signed.

Reply:
> Above where?

https://github.com/bitcoin/bitcoin/pull/12136/commits/e25f290841de2c446a237ad3942a0c2c0427ee62#r184790215

> I'm not necessarily sure that that will always work due to the different data types.

It should. Char vectors are always serialized as compactsize-in-bytes + data-bytes. SerializeToVector first serializes to a vector, and then you serialize that. I can create a small example if you want.

This is still writing the input index as a 4-byte int, rather than as a compactsize, btw.